src/runtime/native.rs

```diff
@@ -40,21 +34,101 @@ struct LoadedScriptData {
+        _world: &mut World,
+        _args: serde_json::Value,
+    ) -> anyhow::Result<serde_json::Value> {
+        anyhow::bail!(
+            "Invalid operation. You may have forgotten to register a custom JsRuntimeOp."
+        );
+    }
+}
+
+impl FromWorld for JsRuntime {
+    fn from_world(world: &mut World) -> Self {
+        // Collect custom ops from the runtime config
+        let default_config = default();
+        let config = world
+            .get_non_send_resource::<JsRuntimeConfig>()
+            .unwrap_or(&default_config);
+        let custom_ops = &config.custom_ops;
+
+        // Collect core ops
+        let mut op_map = get_core_ops();
+
+        // Add custom ops to core ops and warn about conflicts
+        for (op_name, op) in custom_ops.into_iter() {
+            if op_map.insert(op_name, op.clone()).is_some() {
+                warn!(
+                    "Custom op name {op_name} conflicts with core op with \
+                    the same name. Custom op will take precedence."
+                );
+            }
+        }
+
+        // Collect ops into a vector while mapping the op names to it's index in the vector
+        let mut ops: Ops = Vec::with_capacity(op_map.len() + 1);
+        ops.push(Arc::new(InvalidOp)); // The first op is the invalid op called when an op is not found
+        let op_indexes = op_map
+            .into_iter()
+            .map(|(name, op)| {
+                ops.push(op);
+                (name, ops.len() - 1)
+            })
+            .collect::<HashMap<_, _>>();
+
+        // Create reverse lookup so we can get the op name from the index for debugging/logging purposes
+        let op_names = op_indexes
+            .clone()
+            .into_iter()
+            .map(|(k, v)| (v, k))
+            .collect();
+        let op_names = OpNames(op_names);
+
+        // Create the Deno extension
+        let ext = Extension::builder()
+            .ops(vec![op_bevy_mod_js_scripting::decl()])
+            // Insert JS from the registered ops
+            .js(op_indexes
+                .iter()
+                .filter_map(|(&op_name, &op_idx)| ops[op_idx].js().map(|js| (op_name, js)))
+                .collect())
+            .build();
+
+        // Create the runtime
         let mut runtime = DenoJsRuntime::new(RuntimeOptions {
-            extensions: vec![
-                ecs::extension(),
-                log::extension(),
-                Extension::builder()
-                    .js(include_js_files!(prefix "bevy", "./native_setup.js",))
-                    .build(),
-            ],
+            extensions: vec![ext],
             ..Default::default()
         });
 
+        // Run our initialization script, inserting the op name mapping into the script before running it
+        let op_map_json = serde_json::to_string(&op_indexes).unwrap();
+        let init_script_src = &include_str!("./js/native_setup.js")
+            .replace("__OP_NAME_MAP_PLACEHOLDER__", &op_map_json);
```

Comment:
Yeah, we might be able to evaluate the JS and receive a function that we could pass our JSON object into to avoid the string substitution, but I feel like it wouldn't be gaining us much and it would make it harder to read.

Since we are statically including the `native_setup.js` at compile time I think it's safe enough.